src/librbd/api/Image.cc

```diff
@@ -158,7 +160,50 @@ int Image<I>::deep_copy(I *src, librados::IoCtx& dest_md_ctx,
     return -ENOSYS;
   }
 
-  int r = create(dest_md_ctx, destname, "", src_size, opts, "", "", false);
+  int r;
+  if (src->parent_md.spec.pool_id == -1) {
+    r = create(dest_md_ctx, destname, "", src_size, opts, "", "", false);
+  } else {
+    std::string snap_name;
+    src->parent->snap_lock.get_read();
+    r = src->parent->get_snap_name(src->parent_md.spec.snap_id, &snap_name);
```

Comment:
Nit: ```src->parent``` might be ```nullptr``` so resolve the snap name after opening the parent image below and then just run ```src_parent_image_ctx->state->snap_set(XYZ)```. In a perfect world we should be able to internally set the snap by id.